.github/scripts/ci/find_broken_wikilinks.py

```diff
@@ -0,0 +1,173 @@
+            return True
+        else:
+            # may have a redirect
+            return check_redirect(f"{directory}/{link}")
+    else:
+        # some other link; don't care
+        return True
+
+def is_in_comment(s, start, end):
+    return s.rfind("<!--", 0, start) != -1 and s.find("-->", end, -1)
+
+def find_link(s, index=0):
+    found_brackets = False
+    started = False
+    start = None
+    mid = None
+    extra = None
+    end = None
+    square_bracket_level = 0
+    parenthesis_level = 0
+    for i, c in enumerate(s[index:]):
+        i += index
+        if not found_brackets and c == '[':
+            if not start:
+                start = i
+                started = True
+            square_bracket_level += 1
+            continue
+        if started and not found_brackets and c == ']':
+            square_bracket_level -= 1
+            if square_bracket_level == 0:
+                if len(s) > i + 1 and s[i + 1] == '(':
+                    found_brackets = True
+                    mid = i + 1
+            continue
+        if found_brackets and (c == ' ' or c == '#' or c == '?'):
+            if extra is None:
+                extra = i
+            continue
+        if found_brackets and c == '(':
+            parenthesis_level += 1
+            continue
+        if found_brackets and c == ')':
+            parenthesis_level -= 1
+            if parenthesis_level == 0:
+                end = i
+                if is_in_comment(s, start, end):
+                    return None
+                if extra is None:
+                    extra = end
+                return {
+                    'whole': s[start:end + 1],
+                    'link': s[mid + 1: extra],
+                    'link+': s[mid + 1: end],
+                    'pos': (start, mid, extra, end)
+                }
+            continue
+    return None
+
+def find_links(s):
+    results = []
+    index = 0
+    match = find_link(s, index)
+    while match:
+        results.append(match)
+        match = find_link(s, match['pos'][3] + 1)
+    return results
+
+def print_error():
+    print(f"{red('Error:')} Some wiki or image links in the files you've changed have errors.\n")
+    print("This can happen in one of the following ways:\n")
+    print("- The article or image that the link points to has since been moved or renamed (make sure to match capitalisation)")
+    print("- The image that the link points to has the wrong extension (must be one of .png, .jpg, or .gif)")
+    print("- The link simply contains typos or formatting errors")
+    print("- The link works, but contains locale selection (e.g. /wiki/en/Article_styling_criteria instead of /wiki/Article_styling_criteria)")
+    print("- The link works, but contains URL-escaped characters (https://en.wikipedia.org/wiki/Percent-encoding). This only applies for links to articles and images inside the wiki.")
+    print("- The link works, but incurs multiple redirects. Use a direct link instead.")
```

Comment:
this is too specific but I don't know how to shorten it or put better, perhaps... (considering we wanted to ban redirects: #2104)

```
The link must not contain explicit locale or be a redirect
```

Reply:
if we are banning redirects, then i can change it to error when it has to follow any redirect (and you'd have to fix all links when moving around articles)

```suggestion
    print("- The link contains explicit locale selection (e.g. /wiki/en/Article_styling_criteria instead of /wiki/Article_styling_criteria)")
    print("- The link contains URL-escaped characters (https://en.wikipedia.org/wiki/Percent-encoding)")
    print("- The link uses redirects")
```

Reply:
on a second thought, this would suck. I think it's better if we agree to prune redirects from time to time

Reply:
There's no need to keep redirects up-to-date (easing that overhead is what they're there for).

Reply:
wdyt about wording?